src/client/datascience/notebook/vscodeNotebookController.ts

```diff
@@ -170,6 +170,13 @@ export class VSCodeNotebookController implements Disposable {
     public async updateNotebookAffinity(notebook: NotebookDocument, affinity: NotebookControllerAffinity) {
         traceInfo(`Setting controller affinity for ${getDisplayPath(notebook.uri)} ${this.id}`);
         this.controller.updateNotebookAffinity(notebook, affinity);
+
+        // Stick in our list of associated documents and indicate controller may have changed.
+        if (!this.associatedDocuments.has(notebook)) {
+            this.associatedDocuments.add(notebook);
+            this._onNotebookControllerSelected.fire({ notebook, controller: this });
```

Comment:
Setting affinity is not the same as setting it as selected.
You could have a notebook that's associated with kernel A, but the recommended kernel (affinity) is Kernel B.

Reply:
I'd consider raising a seprate event and or having a seprate method to get the recommended kernel.
I believe you need this for instances where a kernel isn't seleced, but want to provide the right intellisense, if that's the case a seprate property/event would be better.

Reply:
No this is exactly what I want. 

This is what is happening now.
- User opens a notebook, kernel is associated with document, notebook selection change event fires
- User closes notebook
- User reopens notebook. Nothing fires, controller doesn't even know what the selection is. 

I think what should happen is the same thing as the first one
- User opens a notebook, kernel is associated with document, notebook selection change event fires

Reply:
I need this so that on second open, the intellisense is using the correct (selected) kernel.

Reply:
On first open, the reason it works is because the selection change event fires after setting affinity.

Reply:
This code here:
https://github.com/microsoft/vscode-jupyter/blob/1247ee68b0179fddb828ebb1a104f76ca8807ce5/src/client/datascience/notebook/vscodeNotebookController.ts#L221

Which is triggered by this:
https://github.com/microsoft/vscode-jupyter/blob/1247ee68b0179fddb828ebb1a104f76ca8807ce5/src/client/datascience/notebook/vscodeNotebookController.ts#L144

On a second open that selection changed event doesn't fire.

Reply:
Ok got it, but if you create a brand new notebook, then this event will still get fired even when no kernel is selected.
Today there's a bug in VS Code causing kernels to get auto selecfed, hence this will work.

I'll merge my fix tomorrow into VS Code and then we can re-visit this PR.
I still don't think this is right.

Reply:
Okay trying a different way. We can just remember it like VS code does. When it gets set, stick it in a map.